### PR TITLE
Support setting working directory as ServerOption and cleanup

### PR DESCRIPTION
**Pull request by tv42**

### CODE REVIEW

**Comment by tv42** [comment]
The travis failures seem to be timeouts waiting for the external sftp client to exit. No idea why, but it doesn't seem directly related to my changes.

**Comment by mdlayher** [comment]
I'm in favor, but I do wonder what's up with the CI build.  Maybe @davecheney can shed some more light.

**Comment by mdlayher** [comment]
The integration tests pass with no issues on my local system, Go 1.6 linux/amd64.

**Comment by davecheney** [comment]
I'm sorry I have not had time to review this yet, I have been travelling. Sadly, I'm still travelling, but I'm nearly done.

**Comment by tv42** [comment]
Nudge.

File: server_integration_test.go [hunk 1]
```diff
@@ -358,17 +373,59 @@ func testServer(t *testing.T, useSubsystem bool, readonly bool) (net.Listener, s
...
+	return path, nil
+}
+
 func runSftpClient(t *testing.T, script string, path string, host string, port int) (string, error) {
 	// if sftp client binary is unavailable, skip test
 	if _, err := os.Stat(*testSftpClientBin); err != nil {
 		t.Skip("sftp client binary unavailable")
 	}
+
+	// make a dummy key so we don't rely on ssh-agent
```
Comment (davecheney) [inline server_integration_test.go]: Can you please pull this out into a separate PR.

File: client_integration_test.go [hunk 2]
```diff
@@ -995,17 +996,9 @@ func readHash(t *testing.T, r io.Reader) (string, int64) {
 // writeN writes n bytes of random data to w and returns the
 // hash of that data.
 func writeN(t *testing.T, w io.Writer, n int64) string {
-	rand, err := os.Open("/dev/urandom")
-	if err != nil {
-		t.Fatal(err)
-	}
-	defer rand.Close()
-
 	h := sha1.New()
```
Comment (davecheney) [inline client_integration_test.go]: Can you please pull this change into a separate PR. Thanks

**Comment by davecheney** [comment]
Sorry I lost track of this.

Can you please

1. pull out the unrelated parts of this change into their own PR so they can be merged without being blocked on this PR.
2. Rebase this PR against master

**Comment by eikenb** [comment]
Want to have another go at this? I'm reviewing all the issues and got to #82 which this PR addresses. If you make the changes Dave suggested, pulling out the unrelated changes and rebasing on master, I'll get it merged.

**Comment by progrium** [comment]
Politely pinging @tv42

**Comment by tv42** [comment]
Might take me a while to get going on this again. If someone else wants to pick up the patch, feel free to do so. The changes should be about as self-evident to you as to me, after all this time..

**Comment by eikenb** [comment]
I'm going to go ahead and apply the one commit that makes it so you can run the integration tests without having an ssh-agent running. This seems like a good idea and really has nothing to do with the main point of this PR.

Oh, and I fixed the typo as well.

**Comment by puellanivis** [comment]
Has all the functionality of this patch been extracted into other PRs and/or already committed? If so, we should close this PR.

**Comment by drakkan** [comment]
This feature is now available using the request server and implementing the optional [RealPathFileLister interface](https://github.com/pkg/sftp/blob/master/request-interfaces.go#L93)
